process_settings.go

```diff
@@ -0,0 +1,200 @@
+// Package process_settings implements dynamic settings for the process.
+package process_settings
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/fsnotify/fsnotify"
+	"gopkg.in/yaml.v3"
+)
+
+// A ProcessSettings is a collection of settings files and a target evaluator
+// that can be used to get the value of a settings based on the current targeting.
+type ProcessSettings struct {
+	FilePath            string            // The path to the settings file that was used to create the ProcessSettings
+	Settings            *[]SettingsFile   // The settings files that make up the ProcessSettings
+	TargetEvaluator     TargetEvaluator   // The target evaluator that is used to determine which settings files are applicable
+	Monitor             *fsnotify.Watcher // The file monitor that is used to detect changes to the settings file
+	WhenUpdatedRegistry []func()          // A list of functions to call when the settings are updated
+}
+
+// NewProcessSettingsFromFile creates a new instance of ProcessSettings by
+// loading the settings from a specified file path and using the specified
+// static context to evaluate the targeting.
+func NewProcessSettingsFromFile(filePath string, staticContext map[string]interface{}) (*ProcessSettings, error) {
+	settings, err := loadSettingsFromFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	monitor, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	err = monitor.Add(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProcessSettings{
+		FilePath:        filePath,
+		Settings:        settings,
+		TargetEvaluator: TargetEvaluator{staticContext},
+		Monitor:         monitor,
+	}, nil
+}
+
+// Get returns the value of a setting based on the current targeting.
+// If the requested setting is not found, an error is returned.
+func (ps *ProcessSettings) Get(settingPath ...interface{}) (interface{}, error) {
+	value := ps.SafeGet(settingPath...)
+
+	if value == nil {
```

Comment:
I don't think that's actually the same. Suppose we had an exception because we couldn't parse the JSON file. We wouldn't want to treat that if the setting were missing. That would be equivalent to the Ruby anti-pattern of `rescue nil`.

If you have offer a `SafeGet`, that could intercept the exact exception where it's missing right?

(Uggh, it's ridiculous in 2023 to have to think about hand-writing exception contracts.)

Reply:
Got the functionality put in to support explicitly set `nil` process settings with 372c2d5070ce9b8ba6aaa08e2bda4ce1c28ac60b. That refactor made it a lot more evident that we should have `SafeGet` call `Get` since `Get` needs to discern between `nil` and `not_set` values. Still have the question regarding the actual need for `SafeGet` since it's just swallowing any errors from `Get` the same way caller code could, but that could be debated later for sure.